File: app/frontend/src/domain/models/Review.ts
/**
 * Modelo de dominio para representar una reseña de establecimiento.
 * Contiene toda la información de una reseña según los requisitos del examen.
 */
export interface Review_Model {
    /** ID único de la reseña en MongoDB */
    id: string;
    /** Nombre del establecimiento reseñado */
    establishment_name: string;
    /** Dirección postal del establecimiento */
    address: string;
    /** Latitud obtenida por geocoding */
    latitude: number;
    /** Longitud obtenida por geocoding */
    longitude: number;
    /** Valoración de 0 a 5 puntos */
    rating: number;
    /** URLs de las imágenes en Cloudinary */
    image_urls: string[];
    /** Email del autor de la reseña */
    author_email: string;
    /** Nombre del autor de la reseña */
    author_name: string;
    /** Token OAuth usado para crear la reseña */
    auth_token?: string;
    /** Fecha y hora de creación */
    created_at: Date;
    /** Fecha y hora de caducidad del token */
    expires_at?: Date;
}

/**
 * Datos para crear una nueva reseña.
 * Solo incluye los campos que el usuario proporciona directamente.
 */
export interface Review_Create_Data {
    /** Nombre del establecimiento */
    establishment_name: string;
    /** Dirección postal para geocodificación */
    address: string;
    /** Valoración de 0 a 5 */
    rating: number;
    /** Archivos de imagen a subir */
    images?: File[];
}

/**
 * Respuesta del servicio de geocodificación.
 */
export interface Geocoding_Result {
    /** Latitud encontrada */
    latitude: number;
    /** Longitud encontrada */
    longitude: number;
    /** Nombre completo de la dirección */
    display_name?: string;
    /** Mensaje de advertencia si se usaron coordenadas por defecto */
    warning?: string;
    /** Indica si las coordenadas son por defecto (fallback) */
    is_default?: boolean;
}

File: app/backend/models/review.py
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime


class ReviewModel(BaseModel):
    """
    Modelo de documento MongoDB para reseñas de establecimientos.
    Representa una reseña completa con toda su información.
    """
    id: str | None = Field(None, alias="_id")
    establishment_name: str = Field(..., description="Nombre del establecimiento reseñado")
    address: str = Field(..., description="Dirección postal del establecimiento")
    latitude: float | None = Field(None, description="Latitud obtenida por geocoding")
    longitude: float | None = Field(None, description="Longitud obtenida por geocoding")
    rating: int = Field(..., ge=0, le=5, description="Valoración de 0 a 5 puntos")
    image_urls: list[str] = Field(default_factory=list, description="URLs de imágenes en Cloudinary")
    author_email: str = Field(..., description="Email del autor de la reseña")
    author_name: str = Field(..., description="Nombre del autor de la reseña")
    auth_token: str = Field(..., description="Token OAuth usado al crear la reseña")
    created_at: datetime = Field(default_factory=datetime.utcnow, description="Fecha de creación")
    expires_at: datetime = Field(..., description="Fecha de caducidad del token")

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={
            "example": {
                "establishment_name": "Casa Lola",
                "address": "Calle Granada 46, Málaga",
                "latitude": 36.7220033,
                "longitude": -4.4189788,
                "rating": 4,
                "image_urls": [
                    "https://res.cloudinary.com/demo/image/upload/v1/reviews/img1.jpg",
                    "https://res.cloudinary.com/demo/image/upload/v1/reviews/img2.jpg"
                ],
                "author_email": "usuario@example.com",
                "author_name": "Juan Pérez",
                "auth_token": "ya29.a0AfH6SMBx...",
                "created_at": "2025-12-08T10:30:00Z",
                "expires_at": "2025-12-08T11:30:00Z"
            }
        }
    )
